### ISSUE TITLE
Cluster configuration Behind a UDP Load Balancing

### ISSUE THREAD

**Issue by iAklis** [wazuh/wazuh#573]
Hello

In my mind, I just need to setup the master as auth (TCP1515) behind LB and make clients as logcollector (UDP1514) behind LB. Although, all status is `connected`, actually the files not `synchronized`.  Can I only sync the client.keys?

All managers keep the same key. But there is still

` ERROR: Error 3010 - Error in cluster client: Could not decrypt message from None`

My master ossec conf
```conf
<cluster>
    <name>wazuh</name>
    <node_name>master</node_name>
    <node_type>master</node_type>
    <key>${cluster_key}</key>
    <interval>1m</interval>
    <port>1516</port>
    <bind_addr>0.0.0.0</bind_addr>
    <nodes>
        <node>${localhost_ip}</node>
        <node>${client_ip}</node>
        <node>${client_ip}</node>
        <node>${client_ip}</node>
    </nodes>
    <hidden>no</hidden>
    <disabled>no</disabled>
  </cluster>
```

My clients ossec conf
```conf
<cluster>
    <name>wazuh</name>
    <node_name>client0x</node_name>
    <node_type>client</node_type>
    <key>${cluster_key}</key>
    <interval>1m</interval>
    <port>1516</port>
    <bind_addr>0.0.0.0</bind_addr>
    <nodes>
        <node>${master_ip}</node>
        <node>${localhost_ip}</node>
    </nodes>
    <hidden>no</hidden>
    <disabled>no</disabled>
  </cluster>
```

```
DIRECTORY="/var/ossec"
NAME="Wazuh"
VERSION="v3.2.1"
REVISION="3220"
```

Best regards,
Aklis Chen

**Comment by mgmacias95** [comment]
Hello @iAklis,

The problem can be caused for two reasons:
* The information is not being sent correctly, so the master receives an incorrect message it can't decrypt.
* The key is not the same for all nodes.

Can you please run the cluster in debug mode and post the output? To do so, stop wazuh and run:
```shellsession
# /var/ossec/bin/wazuh-clusterd -fd
2018-04-27 16:29:46,144 INFO: Cleaning database before starting service...
2018-04-27 16:29:47,197 INFO: Starting cluster wazuh
2018-04-27 16:29:47,198 INFO: Listening on port 1516.
2018-04-27 16:29:47,198 INFO: 4 nodes found in configuration
2018-04-27 16:29:47,198 INFO: Synchronization interval: 1m
2018-04-27 16:29:47,202 DEBUG: Elements in requests queue: [('192.168.56.105', False), ('192.168.56.106', False), ('192.168.56.101', False)]
2018-04-27 16:29:47,202 DEBUG: Crontab: starting to sync
2018-04-27 16:29:47,296 DEBUG: CLIENT: Sending 104 of 104
2018-04-27 16:29:47,383 DEBUG: CLIENT: Sending 104 of 104
2018-04-27 16:29:47,467 DEBUG: CLIENT: Sending 104 of 104
2018-04-27 16:29:47,554 INFO: Starting to sync node01's files
2018-04-27 16:29:47,555 DEBUG: Connected to cluster database socket
2018-04-27 16:29:47,555 DEBUG: Nodes to sync: ['192.168.56.101', '192.168.56.105', '192.168.56.106']
2018-04-27 16:29:47,555 INFO: Found 3 connected nodes
```
This will help me to know exactly where is the error taking place.

Also, does `cluster_control` script work for you? You can try the following command, which will send a request to all cluster nodes in order to retrieve their data:
```shellsession
# /var/ossec/bin/cluster_control -n
-----------------------------------------
Node    Address         Type    Status
-----------------------------------------
node01  192.168.56.102  master  connected
node02  192.168.56.101  client  connected
node03  192.168.56.105  client  connected
node04  192.168.56.106  client  connected
-----------------------------------------
```

Best regards,
Marta

**Comment by iAklis** [comment]
Hi @mgmacias95 ,


```
[root@client5] /var/ossec/bin$ ./wazuh-clusterd -fd
2018-04-28 12:44:06,917 INFO: Cleaning database before starting service...
Traceback (most recent call last):
  File "./wazuh-clusterd", line 385, in <module>
    clear_file_status()
  File "./../framework/wazuh/cluster.py", line 668, in clear_file_status
    filter(lambda x: x != '', file_status.split(' ')))}
  File "./../framework/wazuh/cluster.py", line 666, in <dictcomp>
    db_items = {filename:{'md5': md5, 'timestamp': timestamp} for filename,
ValueError: need more than 2 values to unpack
```

Best regards,
Aklis Chen

**Comment by mgmacias95** [comment]
Hello @iAklis ,

`cluster_control -n` only shows the status of each node in the cluster. If all share the correct key and they're working, all should appear as connected in the master node. 

What you're showing is the output of `cluster_control -a`, more precisely, these agents are not being correctly synchronized. You can try using __`cluster_control -f` in all your client nodes__ to force the synchronization. After that, all agents should appear as connected in `cluster_control -a`.

Best regards,
Marta

**Comment by iAklis** [comment]
Hello @mgmacias95, 

After using `cluster_control -f`， the cluster process of master will be crashed. 

 I have to modify  `framework/wazuh/cluster.json`, removing the the `agent-info` and the `agent-groups` and make them synchronous with rsync.

**Comment by mgmacias95** [comment]
Hello @iAklis,

Can you send me the errors you got using `cluster_control -f`? If you had to remove both `agent-groups` and `agent-info` from `cluster.json` file, this process crashed on client nodes right?

Best regards,
Marta

**Comment by mgmacias95** [comment]
Hello again @iAklis,

I'd like to say that we're currently working on a new cluster that will solve all those problems at synchronization time. 

Best regards,
Marta

**Comment by iAklis** [comment]
Hello @mgmacias95 ,

Thank you for your reply, 
 I got the `invalid literal for int() with base 10: 'Command OK'` that was one of several incorrent type conversions.  And the cluster.log released connection reset by peer about the socket. 

After I removed both `agent-groups` and `agent-info` from `cluster.json` , I deleted the `var/db/cluster.db`  before restart the wazhu. It works well.   

Best regards,
Aklis Chen

**Comment by mgmacias95** [comment]
Hello @iAklis,

I'm glad that fix worked for you. The error you're describing is an error from the `wazuh-clusterd-internal` daemon which has been removed in the new version of the cluster (which will be available soon). We can close the issue if you're ok with this fix but I'd like to help you since using `rsync` is not the best option to synchronize your files:

* The master must keep the `agent-info` file with the most recent modification date.
* The clients must keep the `agent-groups` file with the most recent modification date. Usually, the clients keep the `agent-groups` file from the master, but in `cluster.json` is set in both (`"source": "all"`) directions since these files are created in client nodes first (when the agent reports in the client node).
https://github.com/wazuh/wazuh/blob/642556ba02f209e0136bc98df749e2bc3b739f03/framework/wazuh/cluster.json#L38-L48

`rsync` doesn't check this, and it can affect your cluster.

Best regards,
Marta

**Comment by mgmacias95** [comment]
Hello @iAklis,

We have officially released Wazuh v3.2.3 which solves many problems the cluster had. Check our docs to update: https://documentation.wazuh.com/current/user-manual/manager/wazuh-cluster.html#upgrading-from-older-versions

Best regards,
Marta

**Comment by mgmacias95** [comment]
Hello @iAklis,

It's been almost a month since last comment (excluding mine) so I'm closing this issue.

Don't hesitate to create a new one if you still have questions or problems.

Thanks,
Marta
